Don't autoremember flags

smlance opened the pull request:
Based on https://github.com/bundler/bundler/pull/3744.

smlance commented:
(Waiting for https://github.com/bundler/bundler/pull/3939 to be merged.)

smlance commented:
It looks like certain matchers in `matchers.rb`, e.g., `should_be_installed`, rely on settings configured in previous invocations of `bundle`. This makes certain specs that use `bundle` commands with flags fail, because these settings are not persisted to the matcher commands. Thoughts about how to deal with this? @indirect, @segiddins?

segiddins commented:
@smlance allow passing flags to `should_be_installed`?

indirect commented:
I think it makes sense to allow `path` as an argument to `should_be_installed`... is there anything else we need to support? As an alternative to an argument, we could also potentially create a more generic block form for config settings, something like `config(:path => "/wherever") { should_be_installed("rack 1.0") }`

segiddins commented on lib/bundler/settings.rb:
```diff
@@ -176,11 +185,14 @@ def groups_array(group_type, scope)
     # Always returns an absolute path to the bundle directory
     # TODO: Refactor this method
     def path
+      # TODO: Shoudl we consider the @current_config here?
```
why should / shouldn't we?

smlance replied:
Never mind -- I don't think we need to. The comment just below it explains why: `--path` is no longer used, so `bundle install --path foo` or any command that previously accepted `--path` would now be set with `bundle config`, which only set options in the `local` and `global` scopes.

segiddins commented on lib/bundler/settings.rb:
```diff
@@ -260,8 +274,22 @@ def get_array(key)
+
+      if array
+        value = (array.empty? ? nil : array.join(":"))
+        if scope == :current
```
please use a `case` statement for this, and do the scope checking in there

segiddins commented on spec/commands/clean_spec.rb:
```diff
@@ -470,12 +472,19 @@ def should_not_have_gems(*gems)
-    expect(out).not_to include("foo (1.0)")
-    expect(out).to include("rack (1.0.0)")
+    # See cli/install.rb:L76.
+    # FIXME: [user-unfriendly] `clean` only happens if path is set to
```
this looks like a bug that needs to get fixed

segiddins commented on spec/commands/config_spec.rb:
```diff
@@ -14,9 +14,12 @@
-      expect(bundled_app(".bundle")).not_to exist
-      expect(tmp("foo/bar/config")).to exist
-      should_be_installed "rack 1.0.0"
+      # See cli/install.rb:L86
```
please don't lint to lines -- use methods, which are less likely to go out of date

segiddins commented on lib/bundler/settings.rb:
```diff
@@ -9,15 +9,15 @@ class Settings
-      @local_config  = load_config(local_config_file)
-      @global_config = load_config(global_config_file)
+      @root           = root
+      @current_config ||= {}
```
no need to `||=` in `initialize`

segiddins commented on lib/bundler/settings.rb:
```diff
@@ -260,8 +289,20 @@ def get_array(key)
-      self[key] = (array.empty? ? nil : array.join(":")) if array
+    def set_array(key, array, scope)
+      unless [:current, :local, :global].include? scope
+        Bundler.ui.error "Invalid scope #{scope} given. Please use :local or :global."
```
raise an exception?

segiddins commented on lib/bundler/settings.rb:
```diff
@@ -160,11 +185,13 @@ def with(scope = nil)
+      when :local
         @local_config[key] ? @local_config[key].split(" ").map(&:to_sym) : []
       else
         Bundler.ui.error "Invalid scope #{scope} given. Please use :local or :global."
```
raise an exception here?

segiddins commented on spec/support/builders.rb:
```diff
@@ -365,7 +365,6 @@ def build_with(builder, name, args, &blk)
       spec     = nil
 
-      # TODO: [2.0] Change this, since we won't use options[:path]
       options[:path] ||= @_build_path
```
shouldn't we change this?

smlance replied:
Hm. I'll look into this later.

smlance replied:
No. It looks like this options hash is specific to this file and has nothing to do with the options hash passed to `bundle` commands.

segiddins replied:
ok, so the `TODO` was incorrect?

smlance replied:
Yes. I added it incorrectly.

segiddins commented on spec/support/helpers.rb:
```diff
@@ -1,5 +1,13 @@
+      if block_given?
+        config.each {|k, v| Bundler.settings.set_local(k, v) }
+        yield
+        config.each {|k, _| Bundler.settings.set_local(k, nil) }
```
shouldn't this restore the settings?

smlance replied:
Oh -- good point.

indirect commented on lib/bundler/cli/config.rb:
```diff
@@ -109,7 +112,17 @@ def run
+        delete_config("system")
+        # We still want `path` to be set.
+        :no_conflict
+      elsif name == "path.system" and Bundler.settings[:path] and new_value == "true"
```
Oops. We wound up with `path`, `system`, and `path.system`. I intended us to only have two. :) I think `path` and `path.system` is the clearest. Let's get rid of just plain `system`.

indirect commented on spec/support/helpers.rb:
```diff
@@ -1,5 +1,18 @@
 module Spec
   module Helpers
+    def set_temp_config(config)
```
I think renaming this to `with_config` would match with Ruby conventions for methods that take a block and then unset after.

indirect commented on spec/realworld/parallel_spec.rb:
```diff
@@ -9,7 +9,9 @@
-    bundle :install, :jobs => 4, :env => { "DEBUG" => "1" }
+    bundle "config jobs 4"
+    # NOTE: The env option below cannot be configured with `bundle "config env debug"`
+    bundle :install, :env => { "DEBUG" => "1" }
```
it's actually possible to just use `bundle install --verbose` here (verbose output contains the thread numbers, which is why we use verbose output in this test)

segiddins commented on lib/bundler/settings.rb:
```diff
@@ -30,13 +31,17 @@ def [](name)
+    alias_method :[]=, :set_current
+
+    def set_local(key, value)
       local_config_file or raise GemfileNotFound, "Could not locate #{SharedHelpers.gemfile_name}"
```
please don't use `or`

smlance replied:
Would `unless` be preferable?

segiddins replied:
Yes

segiddins commented on spec/support/helpers.rb:
```diff
@@ -1,5 +1,18 @@
+          Bundler.settings.set_local(k, v)
+        end
+        yield
+        config.each {|k, _| Bundler.settings.set_local(k, old_config[k]) }
```
this needs to be in an `ensure`

smlance replied:
Which part?

segiddins replied:
Restoring the old value

segiddins commented:
This seems good?

segiddins commented:
@smlance you'll need to squash the `wip` commits out, they don't provide any information

smlance commented:
Now that I removed the flags from the list of `method_option`s, `thor` detects that the flags are invalid before `bundler` does, so the error messages I added won't ever be shown.

Should each command have its own `path` setting? They all shared the same `path` configuration before, and this seemed to only be a problem for `binstubs` (see the fixed commits in https://github.com/smlance/bundler/commit/fe9e4246804ecab4b6a5b98f563371b8358625c8). Based on the specs, this isn't a problem for the other commands, so I'm okay with allowing them to share it.

indirect commented:
This looks good!

indirect commented:
@homu r+

homu commented:
:pushpin: Commit 6bf840d has been approved by `indirect`

<!-- @homu r=indirect 6bf840d5790ea97e9691a80861d11b155595bd06 -->
